### PR TITLE
fix(concourse): recursively parse plan for resources

### PR DESCRIPTION
**Pull request by jaredstehler**
Concourse build plan json is a recursive structure, not well-suited for pojo-mapping. This changes Plan.getResources() to a recursive parser for objects of shape {id, get/put:{name, type}}

Addresses issue: spinnaker/spinnaker#5895

### CODE REVIEW

**Comment by gal-yardeni** [comment]
hi @jervi, since you have experience reviewing concourse PRs, can you please take a look/tag someone else who might be able to help? Thanks!

File: igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/model/Plan.java [hunk 1]
```diff
@@ -16,73 +16,56 @@
...
+    while (fields.hasNext()) {
+      Entry<String, JsonNode> f = fields.next();
 
-    @Nullable private OnSuccess onSuccess;
+      if (f.getValue().isArray()) {
+        for (int i = 0; i < f.getValue().size(); i++) {
+          JsonNode jsonNode = f.getValue().get(i);
+          if (jsonNode.isObject()) {
+            res.addAll(getResources((ObjectNode) jsonNode));
+          }
```
Comment (jervi) [inline igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/model/Plan.java:57]: Nitpick, put could this be replaced with a stream instead?
```suggestion
        StreamSupport.stream(f.getValue().spliterator(), false)
            .filter(JsonNode::isObject)
            .map(jsonNode -> getResources((ObjectNode) jsonNode))
            .forEach(res::addAll);
```

File: igor-web/src/test/groovy/com/netflix/spinnaker/igor/concourse/model/PlanSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,438 @@
+/*
+ * Copyright 2016 Schibsted ASA.
```
[the remark this answers is not in the document]
Comment (jaredstehler) [reply igor-web/src/test/groovy/com/netflix/spinnaker/igor/concourse/model/PlanSpec.groovy:2]: doh! i keep doing this 🤦

File: igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/model/Plan.java (hunk 1 above)
Comment (jaredstehler) [reply igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/model/Plan.java]: done

**Comment by jaredstehler** [comment]
updated per review comments

**Comment by gal-yardeni** [comment]
thanks for reviewing @jervi!
@jaredstehler can you please handle the merge conflict before merging it? thanks!

**Comment by jaredstehler** [comment]
> thanks for reviewing @jervi!
> @jaredstehler can you please handle the merge conflict before merging it? thanks!

done; sorry to have missed the conflict!

**Pull request merged**